crates/bevy_utils/src/lib.rs

```diff
@@ -32,178 +42,180 @@ impl std::hash::BuildHasher for FixedState {
-        HashSet::with_capacity_and_hasher(capacity, RandomState::default())
-    }
-}
+pub type HashSet<K> = hashbrown::HashSet<K, RandomState>;
 
-/// A stable std hash set implementing `aHash`, a high speed keyed hashing algorithm
+/// A stable hash set implementing `aHash`, a high speed keyed hashing algorithm
 /// intended for use in in-memory hashmaps.
 ///
 /// Unlike [`HashSet`] this has an iteration order that only depends on the order
 /// of insertions and deletions and not a random source.
 ///
 /// `aHash` is designed for performance and is NOT cryptographically secure.
-pub type StableHashSet<K> = std::collections::HashSet<K, FixedState>;
-
-impl<K> AHashExt for StableHashSet<K> {
-    /// Creates an empty `StableHashSet` with the specified capacity with `aHash`.
-    ///
-    /// The hash set will be able to hold at least `capacity` elements without
-    /// reallocating. If `capacity` is 0, the hash set will not allocate.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use bevy_utils::{StableHashSet, AHashExt};
-    /// let set: StableHashSet<i32> = StableHashSet::with_capacity(10);
-    /// assert!(set.capacity() >= 10);
-    /// ```
+pub type StableHashSet<K> = hashbrown::HashSet<K, FixedState>;
+
+/// A pre-hashed value of a specific type. Pre-hashing enables memoization of hashes that are expensive to compute.
+/// It also enables accelerated comparisons using [`Hashed::fast_eq`].
+/// See [`PassHash`] and [`PassHasher`] for a "pass through" [`BuildHasher`] and [`Hasher`] implementation
+/// designed to work with [`Hashed`]
+/// See [`PreHashMap`] for a hashmap pre-configured to use [`Hashed`] keys.
+pub struct Hashed<V, H = FixedState> {
+    hash: u64,
+    value: V,
+    marker: PhantomData<H>,
+}
+
+impl<V: Hash, H: BuildHasher + Default> Hashed<V, H> {
+    /// Pre-hashes the given value using the [`BuildHasher`] configured in the [`Hashed`] type.
+    pub fn new(value: V) -> Self {
+        let builder = H::default();
+        let mut hasher = builder.build_hasher();
+        value.hash(&mut hasher);
+        Self {
+            hash: hasher.finish(),
+            value,
+            marker: PhantomData,
+        }
+    }
+
+    /// The pre-computed hash.
+    #[inline]
+    pub fn hash(&self) -> u64 {
+        self.hash
+    }
+}
+
+impl<V, H> Hash for Hashed<V, H> {
+    #[inline]
+    fn hash<R: Hasher>(&self, state: &mut R) {
+        state.write_u64(self.hash);
+    }
+}
+
+impl<V, H> Deref for Hashed<V, H> {
+    type Target = V;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.value
+    }
+}
+
+impl<V: PartialEq, H> Hashed<V, H> {
+    /// A faster version of [`PartialEq`] that first checks that `other`'s pre-computed hash
+    /// matches this value's pre-computed hash. For complex types, this can be much faster than
+    /// [`PartialEq`].
+    #[inline]
+    pub fn fast_eq(&self, other: &Hashed<V, H>) -> bool {
+        self.hash == other.hash && self.value.eq(&other.value)
```

Comment:
Is there a reason why this is not just the default `PartialEq` impl? The invariant for `Hash` vs `Eq` (https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq) should guarantee that this shortcircuit is correct (assuming value doesn't have internal mutability and `Hash` and `PartialEq` are implemented correctly).

Reply:
To avoid the extra comparison on hash collisions. Given that we are using these types to optimize hashed collections, it felt right to optimize for that. But I guess PartialEq comparisons are probably still higher frequency than hash collisions, on average, even for `Hashed<T>` types.

Reply:
For cases where pre-hashing is required, I would assume that the equality check is as expensive as `Hash` on the type, and a singular branch is likely to save time on average. Can't say without a benchmark though.

Another option to mitigate the cost of the branch is to provide a compiler hint that the hash check is almost always false (for properly hashed u64 outputs), though most modern CPUs will learn that once it's used enough anyway.

Reply:
hashbrown will call `eq` on two things with distinct hashes if its bucket strategy results in a collision, this doesn't require a true hash collision. Basically there are only so many buckets and it is bucket collisions that cause `eq` to be called, you don't need the hashes to actually be equal.